src/algebra/fields/bigint.tcc

```diff
@@ -164,6 +165,78 @@ bool bigint<n>::test_bit(const std::size_t bitno) const
     }
 }
 
+template<mp_size_t n> template<mp_size_t m>
+inline bigint<n> bigint<n>::operator+(const bigint<m>& other) const
+{
+    static_assert(n >= m, "first arg must not be smaller than second arg for bigint add");
+    bigint<n> res;
+    mpn_add(res.data, data, n, other.data, m);
+    return res;
+}
+
+template<mp_size_t n> template<mp_size_t m>
+inline void bigint<n>::operator+=(const bigint<m>& other)
+{
+    static_assert(n >= m, "first arg must not be smaller than second arg for bigint in-place add");
+    mpn_add(data, data, n, other.data, m);
+}
+
+template<mp_size_t n> template<mp_size_t m>
+inline bigint<n+m> bigint<n>::operator*(const bigint<m>& other) const
+{
+    static_assert(n >= m, "first arg must not be smaller than second arg for bigint mul");
+    bigint<n+m> res;
+    mpn_mul(res.data, data, n, other.data, m);
+    return res;
+}
+
+template<mp_size_t n> template<mp_size_t d>
+inline void bigint<n>::div_qr(bigint<n-d+1>& quotient, bigint<d>& remainder,
+                              const bigint<n>& dividend, const bigint<d>& divisor)
+{
+    static_assert(n >= d, "dividend must not be smaller than divisor for bigint::div_qr");
+    assert_except(divisor.data[d-1] != 0);
+    mpn_tdiv_qr(quotient.data, remainder.data, 0, dividend.data, n, divisor.data, d);
+}
+
+// Return a copy shortened to m limbs provided it is less than limit, throwing std::domain_error if not in range.
+template<mp_size_t n> template<mp_size_t m>
+inline bigint<m> bigint<n>::shorten(const bigint<m>& q, const char *msg) const
```

Comment:
Is this style of passing a message for the ultimate assertion common? I've never seen it before.